eip_7594/src/lib.rs

```diff
@@ -188,47 +184,82 @@ pub fn get_custody_columns(node_id: NodeId, custody_subnet_count: u64) -> Vec<Co
     result
 }
 
-pub fn compute_extended_matrix(blobs: Vec<CKzgBlob>) -> Result<ExtendedMatrix> {
-    let mut extended_matrix: Vec<CKzgCell> = Vec::new();
-
+/**
+ * Return the full ``ExtendedMatrix``.
+ *
+ * This helper demonstrates the relationship between blobs and ``ExtendedMatrix``.
+ */
+pub fn compute_extended_matrix(
+    blobs: Vec<CKzgBlob>,
+) -> Result<[MatrixEntry; MAX_CELLS_IN_EXTENDED_MATRIX]> {
     let kzg_settings = settings();
 
-    for blob in blobs {
-        let cells = *CKzgCell::compute_cells(&blob, &kzg_settings)?;
-        extended_matrix.extend(cells);
+    let mut extended_matrix: [MatrixEntry; MAX_CELLS_IN_EXTENDED_MATRIX] =
+        core::array::from_fn(|_| MatrixEntry::default());
+    for (blob_index, blob) in blobs.iter().enumerate() {
+        let (cells, proofs) = CKzgCell::compute_cells_and_kzg_proofs(blob, &kzg_settings)?;
+        for (cell_index, (cell, proof)) in cells.into_iter().zip(proofs.into_iter()).enumerate() {
+            extended_matrix[blob_index * CELLS_PER_EXT_BLOB + cell_index] = MatrixEntry {
+                cell: try_convert_ckzg_cell_to_cell(&cell)?,
+                kzg_proof: KzgProof::try_from(proof.to_bytes().into_inner())?,
+                row_index: blob_index as u64,
+                column_index: cell_index as u64,
+            };
+        }
     }
 
-    let mut array = [CKzgCell::default(); (MAX_BLOBS_PER_BLOCK * NumberOfColumns::U64) as usize];
-    array.copy_from_slice(&extended_matrix[..]);
-
-    Ok(array)
+    Ok(extended_matrix)
 }
 
-fn recover_matrix(
-    cells_dict: &HashMap<(BlobIndex, CellID), CKzgCell>,
+/**
+ * Return the recovered extended matrix.
+ *
+ * This helper demonstrates how to apply ``recover_cells_and_kzg_proofs``.
+ */
+pub fn recover_matrix(
+    partial_matrix: Vec<MatrixEntry>,
     blob_count: usize,
-) -> Result<ExtendedMatrix> {
+) -> Result<[MatrixEntry; MAX_CELLS_IN_EXTENDED_MATRIX]> {
     let kzg_settings = settings();
 
-    let mut extended_matrix = Vec::new();
+    let mut extended_matrix: [MatrixEntry; MAX_CELLS_IN_EXTENDED_MATRIX] =
+        core::array::from_fn(|_| MatrixEntry::default());
     for blob_index in 0..blob_count {
-        let mut cell_ids = Vec::new();
-        for &(b_index, cell_id) in cells_dict.keys() {
-            if b_index == blob_index as u64 {
-                cell_ids.push(cell_id);
+        let mut cell_indexs = vec![];
+        let mut cells = vec![];
+        let mut proofs = vec![];
```

Comment:
you know the size of these and hence can pre-allocated?

Reply:
looking into this

Reply:
not a constant size, but they MUST have the same size